STP - rewrite of gameplay library

oswinso opened the pull request:
## Description
This PR contains the beginnings of the rewrite of the current gameplay library, based off of the design documents written by @HussainGynai [here](https://docs.google.com/document/d/1pf-aaHnvonTPC6ccI7RdHudpuev84fZslMVDq8lw59s/edit).

This PR contains two things:
- `stp` - A library for the STP(A) system.
- `rj_gameplay` - The actual application level code that uses the `stp` library.

In addition, this PR also adds a few things:
- Tests for `stp` using pytest
- Custom `pylint` linter to check that we don't do `from ... import ...` in any of our code as it breaks hot module reloading.
- Coverage using codecov

oswinso committed:
Added gameplay rewrite + make it installable.

oswinso committed:
Renamed sheen to stp. Fixed test.

kylestach commented on rj_gameplay/stp/play/__init__.py:
```diff
@@ -0,0 +1,140 @@
+
+
+class TacticsEnumMeta(type):
+    """ Metaclass for
```
???

oswinso replied:
lmao nice.

kylestach commented on rj_gameplay/stp/play/__init__.py:
```diff
@@ -0,0 +1,140 @@
+    """ An entry in the TacticEnum for a play.
+    """
+
+    INVALID_IDX = -1
```
I'm not a fan of sentinels, is there any engineering reason this can't explicitly be `None`?

oswinso replied:
Sure.

kylestach commented on rj_gameplay/stp/game_state.py:
```diff
@@ -0,0 +1,61 @@
+        return "Ball(pos:{}, vel:{})".format(self.pos, self.vel)
+
+
+class GameState:
```
Nit: this is called `WorldState` C++-side.

kylestach commented on rj_gameplay/stp/play/offensive_triangle.py:
```diff
@@ -0,0 +1,20 @@
+    """
+
+    def __init__(self, tactics_factory: tactic.Factory):
+        super().__init__(Tactics(tactics_factory))
```
Clean, damn.

kylestach commented on rj_gameplay/stp/play/pure_play.py:
```diff
@@ -0,0 +1,43 @@
+
+    def tick(self) -> List[skill.ISkill]:
+        role_requests = self.collect_role_requests()
+
```
`TODO` here for actually ticking the tactics and returning their actions?

oswinso replied:
Implemented.

kylestach commented on rj_gameplay/stp/role/__init__.py:
```diff
@@ -0,0 +1,125 @@
+
+
+# (robot: Robot, prev_robot: Robot) -> cost.
+CostFn = Callable[[game_state.Robot, Optional[game_state.Robot]], float]
```
`CostFn` should return a tuple of `(float, bool)` so we can express hard constraints.

kylestach replied:
Also, should probably accept a `WorldState`.

oswinso replied:
Yeah true.

HussainGynai replied:
I thought cost functions were going to be different from the hard constraints

kylestach commented on rj_gameplay/stp/role/cost.py:
```diff
@@ -0,0 +1,72 @@
+        # Add the cost for switching robots.
+        cost += flat_switch_cost(robot, prev_robot, switch_cost)
+
+        return max(min(cost, 1.0), 0.0)
```
Are we intending to limit all of our costs to the (0, 1) range? If so, this should happen in a "soft" way (i.e. by tweaking constants so it happens like that in normal operating conditions) rather than by hard-limiting. This just loses signal and makes it more confusing to combine things later.

oswinso replied:
I think it would be nice to limit all of our costs to [0, 1], just so it's easier to compare costs or something. I'll need to think about it more about what the best way of doing this is.

kylestach replied:
I personally really hate limiting costs - it tends to just mask over the weird edge cases. I would prefer to express all costs in (approximate) seconds or something, that tends to make it easier to reason about.

oswinso replied:
What do you mean by (approximate) seconds?

kylestach replied:
I mean like we should scale them to a physically intuitive unit. Seconds was my first thought, because it's pretty easy to define and reason about.
My preferred definition of cost would be, roughly, "time until the robot is able to start actively participating in this role". For example, for a capture skill, this would equate to "time to capture the ball". In this context, I used "approximate" because obviously the cost function won't know _exactly_ how long it is.

kylestach replied:
Bump on this?

oswinso replied:
Done.

kylestach commented on rj_gameplay/stp/role/assignment/naive.py:
```diff
@@ -0,0 +1,164 @@
+        game_state: stp.game_state.GameState,
+        prev_results: assignment.FlatRoleResults,
+    ) -> np.ndarray:
+        """Computes the m x n cost matrix corresponding to the passed in free robots and
```
What are `m` and `n`? (I know from reading the code, but put it in the docstring)

oswinso replied:
Done.

kylestach commented on rj_gameplay/stp/role/assignment/naive.py:
```diff
@@ -0,0 +1,164 @@
+            for robot_idx, robot in enumerate(free_robots):
+                # Otherwise, record the cost.
+                cost: float = request.cost_fn(robot, prev_results.get(role_id, None))
+                if not isfinite(cost):
```
I think we were planning to express hard constraints as a separate `bool` return value. Infinities are annoying because inevitably someone will try to combine two costs and forget to deal with that case, making it explicit makes it impossible to express.

If there is a `NaN` in the cost function, role assignment should be noisy and throw an exception.

That said, is a `0` the right thing to do in this case? Because that's what you're currently putting into the cost matrix, which seems _really_ wrong.

oswinso replied:
The `isfinite` here was mainly to handle bad cost functions, ie trying to divide by 0 or something, rather than hard constraints. This still needs to be worked on, but I just thought that it would be nice if role assignment still worked, but it gets blacklisted + a log message or something.

kylestach replied:
Ok. Yeah honestly I think we should just throw for now.

kylestach commented on rj_gameplay/stp/skill/__init__.py:
```diff
@@ -0,0 +1,77 @@
+SkillT = TypeVar("SkillT", bound=ISkill)
+
+
+class Registry(MutableMapping):
```
`SkillRegistry`?

oswinso replied:
I was thinking about this, and because we're only doing `import .. as ...` calls, it's basically the difference between doing `skill.Registry` vs `skill.SkillRegistry`. I thought that the former was nice, but idk.

kylestach replied:
True...I am 90% sure someone will `from skill import Registry` at some point, though, and I'd rather not constantly police for that.

oswinso replied:
That breaks hot reload though... I think maybe I'll make a custom pylint checker that errors on any `from stp.* import .*` if it's not a test.

kylestach replied:
Oh shit, you right. Yeah that sounds like a better path.

kylestach commented on rj_gameplay/stp/skill/__init__.py:
```diff
@@ -0,0 +1,77 @@
+        return self._dict.__contains__(item)
+
+
+class Factory:
```
`SkillFactory`?

kylestach commented on rj_gameplay/stp/skill/ball_carrier.py:
```diff
@@ -0,0 +1,37 @@
+
+class IBallCarrier(skill.ISkill, ABC):
+    @abstractmethod
+    def get_pass(self) -> Optional[Pass]:
```
I'm pretty sure this belongs at the tactic level. The functionality exposed should be `do_pass(self, pass: Pass)` and `do_shoot(self, shot: Shot)`.

oswinso replied:
This isn't really meant to be an actual working play, but more as like something to think about how the library should look like in terms of ergonomics and stuff.

oswinso replied:
How would `do_pass` and `do_shoot` work? Why does it belong at the tactic level?

kylestach commented on rj_gameplay/stp/skill/ball_carrier.py:
```diff
@@ -0,0 +1,37 @@
+    def __init__(self, passer: int, receiver: int, pt: np.ndarray):
+        self.passer = passer
+        self.receiver = receiver
+        self.pt = pt
```
This should be somewhere else. Also, please specify `pt`'s dimensions.

Also, pass should contain `velocity` and `pass_start_time`.

oswinso replied:
This isn't really meant to be an actual working play, but more as like something to think about how the library should look like in terms of ergonomics and stuff.

kylestach commented on rj_gameplay/stp/skill/ball_carrier.py:
```diff
@@ -0,0 +1,37 @@
+    receiver: int
+    pt: np.ndarray
+
+    def __init__(self, passer: int, receiver: int, pt: np.ndarray):
```
Can we make a `RobotId` alias for `int`? Been meaning to do that C++-side.

oswinso replied:
Done

kylestach commented on rj_gameplay/stp/utils/edict.py:
```diff
@@ -0,0 +1,65 @@
+        return self.concrete_cls
+
+
+class EDict(Generic[ValueInterfaceT], MutableMapping):
```
Honestly I have no idea what the `E` in the name here is intending to convey.

oswinso replied:
I kind of needed a name for this but I couldn't think of anything good.

kylestach replied:
Naming things is hard...maybe `TypedKeyDict`? Still a little bit weird but I think it gets the point across a little better.

oswinso replied:
Sure.

kylestach commented on rj_gameplay/stp/tactic/__init__.py:
```diff
@@ -0,0 +1,182 @@
+class ITactic(ABC):
+    @abstractmethod
+    def get_requests(self, prev_skills: SkillsDict) -> RoleRequests:
+        ...
```
`TODO` for a `tick` method?

kylestach commented on rj_gameplay/stp/tactic/def_support.py:
```diff
@@ -0,0 +1,25 @@
+
+        self.STUB = self.skills.STUB
+
+    def define(self, prev_skills: tactic.SkillsDict) -> tactic.RoleRequests:
```
Wrong name?

kylestach commented on rj_gameplay/stp/tactic/pass_or_shoot.py:
```diff
@@ -0,0 +1,48 @@
+            role_requests[self.RECEIVER] = self.RECEIVER.skill.create_request()
+
+        role_requests[self.BALL_CARRIER] = self.BALL_CARRIER.skill.create_request()
+        role_requests[self.SEEKERS] = self.SEEKERS.skill.create_request()
```
How do we support multiple (an arbitrary number of) seekers in this paradigm?

oswinso replied:
This is still WIP but I was thinking that the `RoleRequest` has a way of specifying the min / max number of robots it is accepting or something. So the `RoleRequest` that `self.SEEKERS.skill.create_request` returns will allow for accepting multiple robots.

kylestach replied:
Why not just create multiple role requests?

oswinso replied:
That could work as well. We would need to know how many robots there are though, so I guess we can pass world_state in here.

oswinso replied:
That could work as well. We would need to know how many robots there are though, so I guess we can pass world_state in here.

kylestach replied:
I think the original plan was that you just request as many roles as you can possibly use (but at lower priority), and then check if they're filled in the `tick` function, and then the roles are just stored as member variables. Is there a reason you moved away from this? (I don't love it given that it introduces additional statefullness, so if you have better ideas I'd love to hear them)

oswinso replied:
That's what I meant by the "we would need to know how many robots there are" part, as in like you need to know how many roles to make.

kylestach replied:
Sure, the idea was just that we request more roles than will be fillable and you only fill some of them. But sure, the role request function can accept a world state.

oswinso replied:
Yeah, but like how many roles do you actually request? 5? 10? 15? That's dependent on how many robots you actually have right?

kylestach reviewed (changes requested):
First pass. Looks mostly pretty clean, although I'm not sure how the dict-of-skills paradigm you've chosen works out:
 - What if you want arbitrarily many (i.e. seeker) skills?
 - How will the play keep track of which tactic gets which `prev_skills` (and, when `tick()` is implemented, which tactic gets which `assigned_skills`)?

oswinso committed:
Created stp.utils.enum to reduce duplicated code.
Moved actual application code to rj_gameplay.
Renamed game_state.py -> rc.py, GameState -> WorldState.

oswinso committed:
edict -> typed_key_dict

oswinso committed:
Changed tactics to tick skills when ticked, returning actions instead of skills.

oswinso committed:
Changed to ignore all .egg-info.

oswinso committed:
Fixed a lot of pylint warnings.

oswinso committed:
Created custom pylint checker that checks for "from ... import ..." as it breaks hot reload.

oswinso committed:
Changed tactic.RoleRequests from Dict[SkillEntry, role.RoleRequest] -> Dict[SkillEntry, List[role.RoleRequest]].

oswinso committed:
Created test_flatten_requests, fixed bug with __str__ for SkillEntry.

oswinso committed:
Add pylint to CI + try out codecov

oswinso committed:
python -> python3

oswinso committed:
Updated gitignore + CI

oswinso committed:
Fixed mypy errors

oswinso committed:
python -> python3

oswinso committed:
Fixed test_cost

oswinso committed:
Added isort

oswinso committed:
Fixed missing cov-report

oswinso committed:
Added situation analyzer + tests.

oswinso committed:
Created .coveragerc to omit test directories.

oswinso committed:
Implemented Coordinator.

oswinso committed:
Implemented hard constraints.

oswinso committed:
Fix lint + mypy.

kylestach commented on rj_gameplay/rj_gameplay/tactic/pass_or_shoot.py:
```diff
@@ -0,0 +1,78 @@
+
+
+@dataclass
+class Props:
```
Move this inside `PassOrShoot`

kylestach commented on rj_gameplay/rj_gameplay/tactic/pass_or_shoot.py:
```diff
@@ -0,0 +1,78 @@
+class PassOrShoot(tactic.ITactic[Props]):
+    """Tactic that controls one ball carrier and multiple seekers."""
+
+    __slots__ = ["skills", "BALL_CARRIER", "RECEIVER", "SEEKERS"]
```
Do we really need to have this everywhere? Kinda a premature optimization.

oswinso replied:
I'm mainly using it as "documentation, in that it's useful to be able to tell all the instance attributes that a class has and have the IDE / interpreter enforce that for me.

kylestach replied:
Sure, fine with me.

kylestach commented on rj_gameplay/tests/stp/utils/test_edict.py:
```diff
@@ -0,0 +1,169 @@
+from typing import Type
```
This file should be renamed

oswinso replied:
Done.

kylestach commented on rj_gameplay/stp/situation/analyzer.py:
```diff
@@ -0,0 +1,67 @@
+
+        return highest_score_situation, situation_play
+
+    def _applicable_situations(
```
We usually consider situation analysis as a function from WorldState -> Situation, not to [Situation]. I'm not opposed to changing it - I think this could be a more flexible model - but that's something that needs to be brought up to the team.

kylestach commented on rj_gameplay/stp/rc.py:
```diff
@@ -0,0 +1,64 @@
+RobotId = int
+
+
+class Robot:
```
We should make it explicit that this is the state of the robot and not a place to apply actions to the robot.

oswinso replied:
Done.

kylestach reviewed (changes requested):
Nits. Some more major things from the first review are still unresolved.

oswinso committed:
Reverted situation analysis to use the old system.

oswinso committed:
Changed Robot and Ball to be read only.

oswinso committed:
Fixed pylint.

oswinso committed:
Removed clamping costs from 0 to 1.

oswinso committed:
Renamed test_edict to test_typed_key_dict.

oswinso committed:
Moved Props for PassOrShoot inside the class.

JNeiger commented:
Imma be honest, I don't have access to the meeting notes/most of the discussion around this; add in all the python framework, shorthand, and magic happening, there's no way for me to actually be able to do more than "yep this is python". And tbf, I'm not sure I'm sold enough on the rewrite to spend a bunch of time trying to reverse engineer dozens of meetings of discussion to figure it out enough to correctly do an in-depth review

I don't mind reviewing code that I've touched or some actual strategy stuff, but I think I'll have to skip most of the new internal framework things.

kylestach commented on rj_gameplay/rj_gameplay/situation/decision_tree/analyzer.py:
```diff
@@ -0,0 +1,197 @@
+        :return: The current BallPos.
+        """
+
+        # TODO: Re-implement all the logic here.
```
Fine for now, could you make an issue?

oswinso replied:
Done.

kylestach commented on rj_gameplay/rj_gameplay/situation/decision_tree/analyzer.py:
```diff
@@ -0,0 +1,197 @@
+            return self.__analyze_normal(world_state, game_info, heuristics)
+
+    @staticmethod
+    def __analyze_restart(
```
Nit: Isn't convention usually one leading underscore, not two?

oswinso replied:
I originally intended this to be a "private" method, but then yeah could be single underscore.

kylestach replied:
Actually, this is fine.

kylestach commented on rj_gameplay/rj_gameplay/situation/decision_tree/plays.py:
```diff
@@ -0,0 +1,168 @@
+"""This module contains all the plays used by the decision tree situation analyzer.
+Ported from the old situation alaysis."""
```
Are these each supposed to contain a list of plays? That seems somewhat backwards. _All_ plays should provide an `is_applicable` method and then we should determine which plays we can use at runtime. Unless I'm misinterpreting this, it prevents us from dynamically selecting playbooks, which is a _very_ important feature.

oswinso replied:
Nah, this is just a replacement for the enum we've been using before, but like it allows us to extend `ISituation` to in the future to require things for all situations, ie. "name" or "description" or whatever. This could be a big enum, but we'd need to do some metaclass magic to do the same thing. Both are identical approaches, I just thought that this was simpler.

kylestach replied:
Sure, this is fine I guess. Could you leave a comment about the intentions (basically just an open enum)?

kylestach commented on rj_gameplay/stp/coordinator.py:
```diff
@@ -0,0 +1,56 @@
+        """
+
+        # Call situational analysis to see which play should be running.
+        cur_situation, cur_play = self._play_selector.select(world_state)
```
This will create a new play instance every tick, which kind of breaks everything. The current logic is, if the situation changes and `is_applicable` no longer applies, you can call `play.try_preempt()` which will either end the play or refuse to give up control (most plays just give up).

https://github.com/RoboJackets/robocup-software/blob/73c92878baf960844b5a4b34c72804093f1ea459/soccer/gameplay/situational_play_selection.py#L340-L348

kylestach replied:
Resolved offline; we're going to try this stateless version to see how it works and if it is bad we'll remove it. For now, no changes needed.

kylestach commented on rj_gameplay/stp/role/cost.py:
```diff
@@ -0,0 +1,83 @@
+import stp.role as role
+
+
+def flat_switch_cost(
```
Previous comments on this still apply.

oswinso replied:
wdym?

kylestach replied:
You marked this comment chain: https://github.com/RoboJackets/robocup-software/pull/1577/files/b8e6f7053e87e223dd7a2a5005eeda4799a7708d#r491723289 as resolved. My concern mentioned is still not addressed (i.e. I don't like trying to saturate costs).

oswinso replied:
You're looking at old code. Check the Files Changed tab.

oswinso committed:
Added issue number.

kylestach commented on rj_gameplay/stp/role/cost.py:
```diff
@@ -0,0 +1,83 @@
+        world_state: rc.WorldState,
+    ):
+        dist: float = np.linalg.norm(robot.pose[:2] - eval_pt)
+        cost: float = min(dist, saturate_dist)
```
(From comment chain above) this still saturates, and we still haven't really defined units for cost which I think would address the original problem.

oswinso replied:
We discussed and said that this was fine.

oswinso committed:
Modified Coordinator, IPlay and PurePlay to actually use handle props.
